src/bin/genemichaels.rs

```diff
@@ -199,7 +237,104 @@ fn main() {
+                eprintln!("\x1B[1;32m  Formatting\x1B[0;22m {}...", if args.folder {
+                    "folder"
+                } else {
+                    "workspace"
+                });
+                let mut workspace_cargo_toml = None;
+                let c_dir = current_dir()?;
+                if args.workspace {
+                    let mut at: Option<&Path> = Some(&c_dir);
+                    while let Some(d) = at.take() {
+                        if d.join("Cargo.toml").exists() &&
+                            cargo_manifest::Manifest::from_path(d.join("Cargo.toml"))
+                                .ok()
+                                .and_then(|m| m.workspace)
+                                .is_some() {
+                            workspace_cargo_toml = Some(d.join("Cargo.toml"));
+                            break;
+                        }
+                        at = d.parent();
+                    }
+                }
+                let (dirs, working_folder) = match workspace_cargo_toml {
+                    Some(ct) => {
+                        // if workspace exists
+                        let path = ct.canonicalize()?;
+                        let workspace_folder = path.parent().unwrap();
+                        let manifest = cargo_manifest::Manifest::from_path(&path)?;
+                        match manifest.workspace.map(|wkspc| wkspc.members) {
+                            Some(mut mb) => {
+                                mb.push(String::new());
+                                (mb, workspace_folder.to_path_buf())
+                            },
+                            // if no members, assume single folder workspace
+                            None => (Vec::from([String::new()]), c_dir),
+                        }
+                    },
+                    // if just checking current folder, same logic as single folder workspace
+                    None if args.folder => {
+                        if !c_dir.join("Cargo.toml").exists() {
+                            return Err(anyhow::anyhow!("No Cargo.toml found in current folder!"));
+                        }
+                        (Vec::from([String::new()]), c_dir)
+                    },
+                    None => {
+                        return Err(anyhow::anyhow!("No workspace found!"));
+                    },
+                };
+                let mut threads = Vec::new();
+                for dir in dirs {
+                    let mut path = working_folder.to_path_buf();
+                    path.push(dir);
+                    for f in walkdir::WalkDir::new(&path) {
+                        match f {
+                            Ok(file) => {
+                                let file_path = file.path().to_path_buf();
+                                if file_path.extension() == Some(OsStr::new("rs")) &&
+                                    !file_path.components().any(|c| c.as_os_str() == OsStr::new("target")) {
```

Comment:
A blacklist approach like this seems risky to me because they could have other dirs they don't want to format -- using the cargo meta was a good idea because it should allow pinpoint precision here.  AFAICT there are 5 directories that can contain source (lib, bin, text, example, bench) and we can explicitly get those dirs from the manifest and use the defaults if they're not present.

Reply:
Excellent idea!